New `ot-tidy.py` script for development and CI

ziuziakowska opened the pull request:
This replaces the existing `ot-tidy.sh` script with a new Python script.

This new script should make it easier to run the linter locally in the same way as CI, as the older wrapper script did not have any usage information.

The script also checks for some misconfigurations and reports those - such as QEMU not being configured to use Clang which results in a different incompatible set of warning flags to be used with GCC to be used for `clang-tidy`.

In addition, this script adds a `-j` flag to run linting in parallel, using the `run-clang-tidy` wrapper that comes with `clang-tidy`. On my machine, this reduced linting time on the same set of files as CI from ~3 minutes to 30 seconds.

jwnrt reviewed (commented):
I get this error from the script:

```console
$ scripts/opentitan/ot_tidy.py --ci-files
Could not parse 'clang-tidy' version string.
```

Here's my output of `clang-tidy --version`:

```console
$ clang-tidy --version
LLVM (http://llvm.org/):
  LLVM version 16.0.2
  Optimized build.
```

I think it's because my `version_line` is the second line, not the first

ziuziakowska commented:
> I think it's because my `version_line` is the second line, not the first

Fixed it to look for that on any line.
Mine appeared like this:
```
$ clang-tidy-20 --version
Ubuntu LLVM version 20.1.2
  Optimized build.
```

jwnrt commented:
Yep, that works now, thanks

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+# SPDX-License-Identifier: Apache2
+
+"""
+'clang-tidy' wrapper script
```
Maybe add a bit more info about the features of this wrapper (and maybe add that clang tidy is the LLVM C linter)?

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+CLANG_MAJOR = 20
+
+# Directory we are in
+ot_scripts = dirname(abspath(__file__))
```
I think all those should be uppercased.
See pylint output

ziuziakowska replied:
What pylint script or wrapper are you using? Most of these issues have not been picked up by the pylint command that CI uses.

rivos-eblot replied:
pylint + flake8. It seems the later is missing from the CI script.
Config file is here: `scripts/opentitan/.flake8`

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+        if entry.is_file() and splitext(entry)[1] == '.lst':
+            lst_files_list.append(entry)
+    for entry in lst_files_list:
+        with open(entry, "r") as f:
```
Prefer "rt" or "rb"

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+    eprint(f"No 'run-clang-tidy-{CLANG_MAJOR}' or suitable 'run-clang-tidy' in PATH.")
+    sys.exit(1)
+
+def main():
```
Not a big deal, but can you try to make the syntax closer to existing scripts/opentitan/*.py scripts?
- try/except exception, with`-d` to show the traceback
- only exit from main, use exception to return error from subroutines
- use argparser.error to report usage error

Please also keep single letter argument options.

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+    if args.ci_files:
+        args.files += ci_files()
+
+    if len(args.files) == 0:
```
`if not args.files`

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+    args.build_dir = abspath(args.build_dir)
+
+    # filter anything that could be interpreted as a flag by clang-tidy
+    args.files = [f for f in args.files if not f.startswith('-')]
```
No sure: can this occur with argparser?

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+
+    if len(args.files) == 0:
+        eprint("No input files provided")
+        sys.exit(1)
```
It would be better to use argparser.error

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+    parser.add_argument('--ci-files', required=False, action='store_true',
+        help="Run on all files checked by CI")
+
+    parser.add_argument('-f', '--files', required=False, nargs="+", action='extend', default=[],
```
80-col limit is likely overshot.

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+
+def eprint(*args, **kwargs):
+    """Wrapper for print to stderr with colour"""
+    col = "\x1b[33m"
```
Beware you need to test for TTY output first, as stderr may be routed to a non-tty file.

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+    col = "\x1b[33m"
+    reset = "\x1b[0m"
+    print(col, end='', file=sys.stderr)
+    print(*args, *kwargs, end='', file=sys.stderr)
```
You likely want to use a formatter here, such as `pprint.pp`

rivos-eblot replied:
`**kwargs`?

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+    print(reset, file=sys.stderr)
+
+
+def check_compile_commands(build_dir):
```
Please add type hinting, _e.g._

`def check_compile_commands(build_dir: str) -> None`

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+    with open(join(build_dir, "config.status"), "r") as f:
+        lines = f.readlines()
+        for line in lines:
+            m = re.match("^CC='(.*)'$", line)
```
I think it is safer to use an r-string: r"^CC...."

rivos-eblot replied:
please prefer multichar variable names

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+    check_cc(args.build_dir)
+
+    cmd = ""
+    cmd_args = []
```
Maybe define the common args here, and only happen the specific args with `.extend()`

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+        """
+    )
+
+    parser.add_argument('-j', '--jobs', required=False, type=uint, nargs='?',
```
I think the default int parser would be enough, so you can test it and return a more explicit error (jobs should be 1 or more) if the value is invalid. Although I think it requires the special `--jobs=-N` to specify a negative integer here.

ziuziakowska replied:
The preferred method seems to be to create a function to use for the `type` parameter. If that returns some kinds of exception then `argparse` prints usage information and a formatted error message. I think it is more consistent to let `argparse` handle the printing of commandline parsing errors.

rivos-eblot replied:
ok

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+    args = parser.parse_args()
+
+    # resolve build dir to absolute path
+    args.build_dir = abspath(args.build_dir)
```
maybe `normpath`?

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+
+    # append ci files
+    if args.ci_files:
+        args.files += ci_files()
```
Prefer `extend()`

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+    chdir(qemu_root)
+    execvp(cmd, cmd_args)
+
+if __name__ == '__main__':
```
As you elected to use " rather than ' for strings you mighht want to tweak the syntax here as well

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+        for line in p.stdout.decode().split("\n"):
+            m = re.match("^.*LLVM version ([0-9]*)\\..*$", line)
+            if m is not None:
+                if m.group(1) != str(CLANG_MAJOR):
```
It is a bit weird to compare integer versions as strings (?)

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+            sys.exit(1)
+        version_found = False
+        for line in p.stdout.decode().split("\n"):
+            m = re.match("^.*LLVM version ([0-9]*)\\..*$", line)
```
Why is is none or more integer? Shouldn't we always match a leading integer here?
(+ _r_-string)

`^.*` is also a bit weird. Should not it be either the start of the line _or_ a space char before LLVM?
I guess it comes from the .sh script:

shell script was using "re.seach(r'LLVM version') then "re.replace(r'^.*LLVM version (\d+)\.d\.\d+.*$', '\1')", which is different that what the new script does.

ziuziakowska replied:
> Why is is none or more integer? Shouldn't we always match a leading integer here? (+ _r_-string)

Changed to `([0-9]+)`.

> `^.*` is also a bit weird. Should not it be either the start of the line _or_ a space char before LLVM? I guess it comes from the .sh script:
> 
> shell script was using "re.seach(r'LLVM version') then "re.replace(r'^._LLVM version (\d+).d.\d+._$', '\1')", which is different that what the new script does.

There seems like there can be arbitrary text before the `LLVM version` string. For example on my machine there is:
```
$ clang-tidy-20 --version
Ubuntu LLVM version 20.1.2
  Optimized build.
```
I think it makes sense to match from the start of the string up to and including the string `LLVM version` and then expect the version to follow after that to the end of the line.

Edit: Meant "from the start of the line"

rivos-eblot replied:
Your are matching one line after another.

ziuziakowska replied:
This is intentional as the first line is not necessarily the one with the version string, such as in @jwnrt's case.

rivos-eblot replied:
I think you can search the whole output with a one liner then:
`re.search(r'(?s)(?:\s|^)LLVM version (\d+)\.', p.stdout)`

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+    # 'run-clang-tidy' does not have a version flag, so return it if it exists
+    if which("run-clang-tidy"):
+        return "run-clang-tidy"
+    eprint(f"No 'run-clang-tidy-{CLANG_MAJOR}' or suitable 'run-clang-tidy' in PATH.")
```
It is a bit weird for the user. I'm not sure what run-clang-tidy is (not installed by default on my system), but the error is hard to tie to the -j N option. May something more explanatory for the user is required: cannot run mulitple jobs without run-clang-tidy tool, or something similar?

ziuziakowska replied:
The [documentation](https://clang.llvm.org/extra/clang-tidy/#running-clang-tidy-in-parallel) claims this script is included with `clang-tidy` and it existed on my system and in CI. Should it error and exit or fallback to regular `clang-tidy`?

rivos-eblot replied:
It guess it depends on how it is installed (package manager, by hand, depending on the OS, ...). I do not have it, with or without -version suffix.
I think in this case, as -j N cannot be satisfied, I think the better is to let the user know that this is not supported, i.e. fail.

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+# clang major version to check for
+CLANG_MAJOR = 20
+
+# Directory we are in
```
nitpick: maybe my understanding is not right, but it is the directory where the script is located, "we" (the user?) is likely in cwd.

rivos-eblot commented on .github/workflows/build_test.yaml:
```diff
@@ -121,7 +124,7 @@ jobs:
         # Expect many warnings/errors (accounted but not reported) from included QEMU files
         run: |
-          scripts/opentitan/ot-tidy.sh --ci -p build-clang
+          scripts/opentitan/ot_tidy.py --build-dir build-clang --ci-files -j
```
can you keep `ot-tidy.py`? Thanks.
rationale: all other scripts in this directory (with a single exception) use `-` as the separator char, and it matches clang- tools.

ziuziakowska replied:
With `ot-tidy.py` as the name I get:
```
scripts/opentitan/ot-tidy.py:1:0: C0103: Module name "ot-tidy" doesn't conform to snake_case naming style (invalid-name)
```

rivos-eblot replied:
You will need to add:

````py
# pylint: disable=invalid-name
# pylint: enable=invalid-name
````
such as in `keymanager-dpe.py` for example.

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+    parser.add_argument('-f', '--files', required=False, nargs="+", action='extend', default=[],
+        help="Files to run tidy on")
+
+    args = parser.parse_args()
```
It think there is a regression here (vs. .sh script): we need to be able to pass any argument to clang-tidy, not just files.

ziuziakowska replied:
Is this used anywhere? I'm not sure how compatible it would be, since `run-clang-tidy` and `clang-tidy` support different flags, and the former doesn't have a way to pass arguments to the latter. Both only have the option to pass additional flags to the compiler with `-extra-arg`/`-extra-arg-before` and `-- [ARGS...]` respectively.

rivos-eblot replied:
Yes we do use it: the public version of OpenTitan emulation is only about a half of what we emulate: we wrote `ot-tidy.sh` so that it could be use by either public OpenTitan or with our extensions. Our CI extensively uses it.

I wonder whether it would make sense to have two tidy helper scripts, as this one is getting complex and is likely to become more difficult to use at least from a developer from the command line. Maybe a dedicated script to run run-clang-tidy would be easier, and keep `ot-tidy.sh` as it is? `ot-tidy.sh` is really though as a very thin wrapper around cland-tidy: it clang-tidy with some sugar syntax.

ziuziakowska replied:
I see. Perhaps it might make more sense to have separate helper scripts then. I originally had the thought to write this because the only usage information was in the CI, and the warning flag errors when not configuring with `--cc=clang` were only documented in the comments there too. The `run-clang-tidy` wrapper also let me lint a lot faster instead of waiting for CI to do it. So I suppose this script is more public development-oriented.

rivos-eblot replied:
out of curiosity: does run-clang-tidy manage output streams from its various subprocesses when several of them report errors?

ziuziakowska replied:
`run-clang-tidy` prints the output generated by a `clang-tidy` subprocess after each one completes, like so:

```
$ scripts/opentitan/ot-tidy.py -j 2 -f hw/opentitan/ot_i2c.c hw/opentitan/ot_otp.c hw/opentitan/ot_aes.c                            ⌛3s
Running clang-tidy for 3 files out of 2598 in compilation database ...
[1/3][0.4s] clang-tidy-20 -p=build -quiet --config-file=scripts/opentitan/clang-tidy.yml /home/alice/qemu/hw/opentitan/ot_otp.c
491 warnings generated.

[2/3][0.9s] clang-tidy-20 -p=build -quiet --config-file=scripts/opentitan/clang-tidy.yml /home/alice/qemu/hw/opentitan/ot_aes.c
../hw/opentitan/ot_aes.c:30:2: error: example error in AES [clang-diagnostic-error]
   30 | #error example error in AES
      |  ^
554 warnings and 1 error generated.
Error while processing /home/alice/qemu/hw/opentitan/ot_aes.c.

[3/3][3.4s] clang-tidy-20 -p=build -quiet --config-file=scripts/opentitan/clang-tidy.yml /home/alice/qemu/hw/opentitan/ot_i2c.c
../hw/opentitan/ot_i2c.c:50:3: error: example warning in I2C [clang-diagnostic-#warnings,-warnings-as-errors]
   50 |  #warning example warning in I2C
      |   ^
```

rivos-eblot replied:
Ok nice to now. Thanks for the feedback.

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+        """
+    )
+
+    parser.add_argument('--ci-files', required=False, action='store_true',
```
I think there is a regression vs the .sh script: how to specify the configuration file?

rivos-eblot commented on docs/opentitan/tools.md:
```diff
@@ -50,7 +50,7 @@ of options and the available features.
   using the same parameters as the KeyManager DPE. It is dedicated to unit test purposes.
 * `ot-format.sh` is a simple shell wrapper to run clang-format (code formatter) on OpenTitan files
-* `ot-tidy.sh` is a simple shell wrapper to run clang-tidy (C linter) on OpenTitan files
+* `ot_tidy.py` is a python wrapper script to run clang-tidy (C linter) on OpenTitan files, optionally in parallel and locally or in CI.
```
line width

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+    cmd_args += args.files
+    chdir(qemu_root)
+    execvp(cmd, cmd_args)
+
```
missing empty line

rivos-eblot replied:
note that this appies to all top level functions in the file.

rivos-eblot reviewed (changes requested):
I think several features are broken, and some are a bit fragile (parsing, ANSI management, ...).

Can you run flake8, as there are many outsized lines and argument aligment warning messages. Thanks.

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+            eprint("'clang-tidy --version' failed.")
+            sys.exit(1)
+        version_found = False
+        for line in p.stdout.decode().split("\n"):
```
I missed this one: use `text=True` in `run` so you get a text stream rather than a byte stream you need to decode

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+                if m.group(1) != str(CLANG_MAJOR):
+                    eprint(f"'clang-tidy' is major version {m.group(1)}, expecting {CLANG_MAJOR}.")
+                    sys.exit(1)
+                version_found = True
```
You can replace version_found with a `else` condition on the `for` loop: if break is not called (i.e. match), then the else condition is executed.

rivos-eblot commented on scripts/opentitan/ot_tidy.py:
```diff
@@ -0,0 +1,209 @@
+    if which(f"clang-tidy-{CLANG_MAJOR}"):
+        return f"clang-tidy-{CLANG_MAJOR}"
+    # Otherwise, look for 'clang-tidy' and check its version
+    if which("clang-tidy"):
```
Note: `which` can be removed here: you can catch `subprocess.run` exception `FileNotFoundError` which indicates that it fails to resolve the path to the application and also catch `CalledProcessError` (with check=True) if `--version` fails.

Both syntax could even be combined (clang-tidy-{CLANG_MAJOR} and clang-tidy) so you could even ensure that a versioned version actually matches its real version (as clang-tidy-{CLANG_MAJOR} are often symlinks)

ziuziakowska committed:
[ot] scripts/opentitan: ot-tidy: create new `ot-tidy.py` script

This script is intendde to replace the `ot-tidy.sh` script for running
clang-tidy on our part of the codebase.

A `-j` flag is added to run clang-tidy jobs in parallel, using the
run-clang-tidy script that comes with it. On my machine this speeds up
linting files for CI from >3 minutes to ~30 seconds.

The script also checks for some configuration issues, such as the C
compiler for QEMU not being set to clang - this causes some warnings
flags to be enabled that are recognised by GCC but not clang, and these
will show up as errors.

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

rivos-eblot commented on scripts/opentitan/ot-tidy.py:
```diff
@@ -0,0 +1,256 @@
+    try:
+        parser = argparse.ArgumentParser()
+
+        parser.add_argument("-b", "--build-dir", required=False,
```
You can remove the `required=False` which is the default behavior for all options

rivos-eblot commented on scripts/opentitan/ot-tidy.py:
```diff
@@ -0,0 +1,256 @@
+                            action="store_true",
+                            help="Run on all files checked by CI")
+
+        parser.add_argument("-f", "--files", required=False,
```
I wonder if this case is allowed: `required=False` means the option is optional, but on the other side `nargs="+"` indicates that the option should be specified at least once.

rivos-eblot replied:
BTW, maybe this which is the main one should use positional argument (`files`) rather than an option?

ziuziakowska replied:
Changed this to be positional with `nargs="*"`

rivos-eblot commented on scripts/opentitan/ot-tidy.py:
```diff
@@ -0,0 +1,256 @@
+        if args.ci_files:
+            args.files.extend(ci_files())
+
+        if not args.files:
```
The option you specify to the parser should handle this case I believe

ziuziakowska replied:
With `files` as a positional argument with `nargs="*"`, this check still needs to be present. It is valid to not provide any files and only use `--ci-files` to append the list of those.

rivos-eblot commented on scripts/opentitan/ot-tidy.py:
```diff
@@ -0,0 +1,256 @@
+            ]
+
+        fmt_args = " ".join(cmd_args[1:])
+        print(f"run '{cmd} {fmt_args} ... ({len(args.files)} files)'")
```
Maybe use a logger for this?

ziuziakowska replied:
I think it is fine to just `print` here to show the command being executed.

rivos-eblot replied:
It departs from the all the other scripts/opentitan tools, and it will be really ugly when --ci-files are used, which ends up printing many lines. Most of the time, I think we only care about the failure, we just need to know what is wrong, don't we? I really think it's nice that when everything's fine, we do not get any output, except maybe a status line.

ziuziakowska replied:
`run-clang-tidy` will already print out how many files it has finished checking/how many there are to check after each one, interspersed with any lint errors, so there will be some output regardless.

I think a `print` to show the main command that the script runs would be useful to see, especially for something user-facing. The file arguments themselves are not printed and show up as `(N files)`.

rivos-eblot replied:
Ok. If you do not want to use a logger, please add a `-v` verbose option so that this command line is not forced-printed out. Thanks

rivos-eblot commented on scripts/opentitan/ot-tidy.py:
```diff
@@ -0,0 +1,256 @@
+    eprint = eprint_tty
+
+
+def check_compile_commands(build_dir) -> None:
```
missing type hint (`: str`)

rivos-eblot commented on scripts/opentitan/ot-tidy.py:
```diff
@@ -0,0 +1,256 @@
+
+eprint = eprint_plain
+if sys.stderr.isatty():
+    eprint = eprint_tty
```
or
````py
eprint = eprint_tty if sys.stderr.isatty() else eprint_plain
````

rivos-eblot commented on scripts/opentitan/ot-tidy.py:
```diff
@@ -0,0 +1,256 @@
+    return [relpath(normpath(join(QEMU_ROOT, f))) for f in ci_files_list]
+
+
+def uint(value) -> int:
```
value: str

rivos-eblot commented on scripts/opentitan/ot-tidy.py:
```diff
@@ -0,0 +1,256 @@
+LLVM_MAJOR = 20
+"""LLVM major version to check for"""
+
+OT_SCRIPTS = dirname(normpath(__file__))
```
my mistake from a previous comment: `abspath` is better than `normpath`, sorry about that.

rivos-eblot commented on scripts/opentitan/ot-tidy.py:
```diff
@@ -0,0 +1,256 @@
+            if match is not None:
+                ccs = [f"clang-{LLVM_MAJOR}", "clang"]
+                if match.group(1) not in ccs:
+                    eprint("QEMU build is configured with non-clang C compiler "
```
May be use `f""` on each line, it should be more readable.
Python would only complain if the resulting string contains no variable.

ziuziakowska committed:
[ot] ci: .github,.gitlab-ci.d: replace `ot-tidy.sh` with `ot-tidy.py`

Replaces all usages and updates docs.

Older `ot-tidy.sh` script is retained for any current users of it.

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

rivos-eblot commented on scripts/opentitan/ot-tidy.py:
```diff
@@ -0,0 +1,252 @@
+    Check for 'compile_commands.json' in given build directory
+    """
+    if not isfile(join(build_dir, "compile_commands.json")):
+        eprint("missing 'compile_commands.json' in build directory "
```
maybe `f"` on the first line?

AlexJones0 commented on scripts/opentitan/ot-tidy.py:
```diff
@@ -0,0 +1,252 @@
+    """Argparse function to parse integer greater than 0"""
+    v = int(value)
+    if v <= 0:
+        raise argparse.ArgumentTypeError("value is an integer less than 0")
```
```suggestion
        raise argparse.ArgumentTypeError("value is an integer <= 0")
```

AlexJones0 commented on scripts/opentitan/ot-tidy.py:
```diff
@@ -0,0 +1,252 @@
+eprint = eprint_tty if sys.stderr.isatty() else eprint_plain
+
+
+def check_compile_commands(build_dir: str) -> None:
```
Nit: not necessary to change for this PR, but my preference would be using the `pathlib.Path` type instead of `str` for path.

AlexJones0 commented on scripts/opentitan/ot-tidy.py:
```diff
@@ -0,0 +1,252 @@
+    cmds = [f"run-clang-tidy-{LLVM_MAJOR}", "run-clang-tidy"]
+    for cmd in cmds:
+        try:
+            _ = subprocess.run([cmd, "--help"], capture_output=True, check=True)
```
Nit - maybe we do and I am just not familiar, but do we need to `capture_output=True` if we immediately discard it? I think you could also query the result's `.returncode` and check that is `== 0`?

ziuziakowska replied:
`capture_output` needs to be enabled to capture the help text, as we do not want to display it.

`subprocess.run` raises an exception if `check=True` and the subprocess exits with a non-zero exit code, so this is already handled.

AlexJones0 replied:
Makes sense, thanks for explaining!

AlexJones0 reviewed (approved):
Thanks, I've left a couple of additional comments. This LGTM though and definitely makes local `clang-tidy` runs faster.

jwnrt commented:
Regression failure is known and unrelated
